File: Engine/src/main/java/org/etieskrill/engine/scene/component/LayoutUtils.java
package org.etieskrill.engine.scene.component;

import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;
import org.joml.Vector2f;
import org.joml.Vector4f;

public class LayoutUtils {

    @Contract("_ -> new")
    static Vector2f getMinNodeSize(Node<?> node) {
        Vector4f margin = node.getMargin();
        return new Vector2f(node.getSize())
                .add(margin.z() + margin.w(), margin.x() + margin.y());
    }

    @Contract("_, _ -> new")
    static Vector2f getPreferredNodePosition(@NotNull Vector2f size, @NotNull Node<?> node) {
        Vector2f nodeSize = new Vector2f(node.getSize()).add(
                new Vector2f(node.getMargin().w(), node.getMargin().y())
                .mul(2)
        );
        
        Vector2f marginPos = new Vector2f(
                node.getMargin().z(),
                node.getMargin().x()
        );

        Vector2f _size = new Vector2f(size);
        if (node.getAlignment() == Node.Alignment.FIXED_POSITION) {
            return node.getPosition();
        }
        Vector2f pos = switch (node.getAlignment()) {
            case FIXED_POSITION -> throw new IllegalStateException();

            case TOP_LEFT -> new Vector2f(0f, 0f);
            case TOP -> _size.sub(nodeSize).mul(0.5f, 0f);
            case TOP_RIGHT -> _size.sub(nodeSize).mul(1f, 0f);
            case CENTER_LEFT -> _size.sub(nodeSize).mul(0f, 0.5f);
            case CENTER -> _size.mul(0.5f).sub(nodeSize.mul(0.5f));
            case CENTER_RIGHT -> _size.sub(nodeSize).mul(1f, 0.5f);
            case BOTTOM_LEFT -> _size.sub(nodeSize).mul(0f, 1f);
            case BOTTOM -> _size.sub(nodeSize).mul(0.5f, 1f);
            case BOTTOM_RIGHT -> _size.sub(nodeSize);
        };
        
        return pos.add(marginPos);
    }
    
}

File: Engine/src/main/java/org/etieskrill/engine/scene/component/layoutUtils.kt
package org.etieskrill.engine.scene.component

import org.etieskrill.engine.scene.component.Node.Alignment.*
import org.joml.Vector2f
import org.joml.Vector2fc
import org.joml.minus
import org.joml.plus
import org.joml.times

internal fun getMinNodeSize(node: Node<*>): Vector2f =
    node.formattedSize + node.margin.run { Vector2f(z + w, x + y) }

internal fun getPreferredNodePosition(size: Vector2fc, node: Node<*>): Vector2f {
    val nodeSize = node.formattedSize + Vector2f(node.margin.w, node.margin.y) * 2f
    val marginPos = Vector2f(node.margin.z, node.margin.x)

    val size = Vector2f(size)
    val pos = when (node.alignment) {
        FIXED_POSITION -> return node.position

        TOP_LEFT -> Vector2f(0f)
        TOP -> (size - nodeSize) * Vector2f(0.5f, 0f)
        TOP_RIGHT -> (size - nodeSize) * Vector2f(1f, 0f)
        CENTER_LEFT -> (size - nodeSize) * Vector2f(0f, 0.5f)
        CENTER -> (size * 0.5f) - (nodeSize * 0.5f)
        CENTER_RIGHT -> (size - nodeSize) * Vector2f(1f, 0.5f)
        BOTTOM_LEFT -> (size - nodeSize) * Vector2f(0f, 1f)
        BOTTOM -> (size - nodeSize) * Vector2f(0.5f, 1f)
        BOTTOM_RIGHT -> size - nodeSize
    }

    return pos + marginPos
}
